op-node/rollup/derive/batches.go

```diff
@@ -25,15 +26,95 @@ const (
 	BatchFuture
 )
 
-func CheckBatchEspresso() BatchValidity {
-	// TODO: verify batch constraints
-	panic("Unimplemented")
+func CheckBatchEspresso(cfg *rollup.Config, log log.Logger, l1Blocks []eth.L1BlockRef, l2SafeHead eth.L2BlockRef, batch *BatchWithL1InclusionBlock, hotshot HotShotContractProvider) BatchValidity {
+	// First, validate the headers that represent the beginning of the L2 block range of this batch.
+	jst := batch.Batch.Justification
+	if jst == nil {
+		// Espresso blocks must have a justification
+		return BatchDrop
+	}
+	startHeaders :=
+		[]espresso.Header{jst.PrevBatchLastBlock, jst.FirstBlock}
+	headerHeights :=
+		[]uint64{jst.FirstBlockNumber - 1, jst.FirstBlockNumber}
+	err := hotshot.verifyHeaders(startHeaders, headerHeights)
+
+	// In the case that the headers aren't available yet (perhaps the validator's L1 client is behind), return BatchFuture so that we can try again later
```

Comment:
We need to think carefully about how this interacts with potential escape hatches. As long as we keep returning `BatchFuture` until the headers are available, this should be fine and deterministic. We may eventually want to return `BatchDrop` if the headers aren't available and the L1 height is ahead of the L1 origin by `sequence_window_size`, or something like that. That would allow the derivation pipeline to continue making progress even if for some reason nobody is posting HotShot commitments to the sequencer contract for a long time.

I don't think we need to worry about this for the demo, but please make an issue with the `op-integration-loose-ends` label so we remember to revisit this if/when we productionize this.

Reply:
added issue #46